src/ragas/testset/evolutions.py

```diff
@@ -1,87 +1,201 @@
         ...
 
-    async def aevolve(self):
+    @abstractmethod
+    async def aevolve(self, llm: BaseRagasLLM, docstore: DocumentStore) -> str:
         ...
 
 
-async def simple_evolution(llm: BaseRagasLLM, seed_doc: Document):
-    human_prompt = SEED_QUESTION.format(context=seed_doc.page_content)
+@dataclass
+class SimpleEvolution(Evolution):
+    def evolve(self, llm: BaseRagasLLM, docstore: DocumentStore):
+        logger.info("evolving question")
+        return asyncio.get_event_loop().run_until_complete(self.aevolve(llm, docstore))
+
+    def _get_more_adjacent_nodes(self, docstore: DocumentStore):
+        """
+        if the evolutions doesn't have enough nodes to frame a question, get more nodes
+        """
+        assert self._root_node is not None, "root node cannot be None"
+        # get more nodes from above the context window
+        prev_adjacent_node = docstore.get_adjacent(self._root_node, Direction.PREV)
+        if prev_adjacent_node is None:
+            # get more nodes from below the context window
+            next_adjacent_node = docstore.get_adjacent(self._root_node, Direction.NEXT)
+            if next_adjacent_node is not None:
+                # add next nodes towards the end
+                self.nodes.append(next_adjacent_node)
+            else:
+                # retry with new base node
+                self.nodes = docstore.get_random_nodes(k=1)
+                self._root_node = self.nodes[0]
+        else:
+            # add prev nodes in index 0
+            self.nodes.insert(0, prev_adjacent_node)
+
+    async def aevolve(self, llm: BaseRagasLLM, docstore: DocumentStore):
+        # can the node be used to frame a question?
+        if self._tries == 0:
+            self.nodes = docstore.get_random_nodes(k=1)
+            self._root_node = self.nodes[0]
+        merged_node = self.merged_nodes()
+        passed, table_is_present = await self.node_filter.afilter(self.nodes[0])
+        if not passed:
+            self.nodes = docstore.get_random_nodes(k=1)
+            return await self.aretry_evolve(llm, docstore, update_count=False)
+
+        # frame a basic question with with node
+        seed_questions = await simple_evolution(llm, merged_node, table_is_present)
+        # NOTE: might need improvement
+        # select only one seed question here
+        seed_question = choice(seed_questions)
+        is_valid_question = await self.question_filter.afilter(seed_question)
+        if not is_valid_question:
+            # get more context to rewrite question
+            self._get_more_adjacent_nodes(docstore)
+            # retry with new nodes added
+            return await self.aretry_evolve(llm, docstore)
+        else:
+            # if valid question
+            return seed_question
+
+
+async def simple_evolution(
+    llm: BaseRagasLLM, seed_doc: Document, is_table_present: bool = False
+):
+    if is_table_present:
```

Comment:
It's better to remove this TABLE_QA part, it was experimental.

Reply:
alrighty - I will remove that logic in the next PR